device_doctor/lib/src/device.dart

```diff
@@ -0,0 +1,98 @@
+// Copyright 2020 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'android_device.dart';
+import 'config.dart';
+import 'ios_device.dart';
+
+/// The root of the API for controlling devices.
+DeviceDiscovery get devices => DeviceDiscovery();
+
+/// Type of host that is configured to test.
+enum HostType { vm, physical }
+
+/// Operating system on the devices that this host is configured to test.
+enum DeviceOperatingSystem { android, ios }
+
+/// Discovers available devices and chooses one to work with.
+abstract class DeviceDiscovery {
```

Comment:
This factory class looks very similar to the checks performed in the config, can we consolidate both on them in a single place?